snapshotstate: add cleanup of abandonded snapshot imports

mvo5 opened the pull request:
When a snapshot import is going on there is a chance that this
gets interrupted because e.g. a reboot is happening or snapd
crashes or restarts. This will leave incomplete import files
around.

This commit will clean those files on startup of snapd and
also adds flock based locks around the importing to make sure
the cleanup is reliable and not racy.

pedronis commented on overlord/snapshotstate/snapshotmgr.go:
```diff
@@ -84,6 +85,15 @@ func (mgr *SnapshotManager) Ensure() error {
 		return mgr.forgetExpiredSnapshots()
 	}
+
+	// cleanup abandonded imports once per startup
```
nowadays this kind of thing can be done from  a Snapshot.StartUp method

stolowski commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -544,13 +568,72 @@ func (t *importTransaction) Cancel() error {
+}
+
+// CleanupAbandondedImports will clean any import that is in progress.
+// This is meant to be call at startup of snapd before any real imports
```
```suggestion
// This is meant to be called at startup of snapd before any real imports
```

stolowski commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -486,34 +492,49 @@ var ErrCannotCancel = errors.New("cannot cancel: import already finished")
+// for the given import file path. It may return an error if an
+// invalid file was specified.
+func newImportTransactionFromImportFile(p string) (*importTransaction, error) {
+	m := regexp.MustCompile("^([0-9]+)_importing$")
```
Perhaps it would make sense to declare all the "...importing" patterns (regexp, file globs) close to each other as const at the top of the package to make sure nothing gets out of sync.

stolowski commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -544,13 +568,72 @@ func (t *importTransaction) Cancel() error {
 // Commit will commit a given transaction
 func (t *importTransaction) Commit() error {
-	if err := os.Remove(importInProgressFilepath(t.id)); err != nil {
+	t.committed = true
```
Probably not very important, but should commited be set to true only if unlock() didn't return an error?

mvo5 replied:
Thanks, I think this makes sense

stolowski commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -544,13 +568,72 @@ func (t *importTransaction) Cancel() error {
+	return t.unlock()
+}
+
+func (t *importTransaction) lock() error {
```
Maybe throw an internal error if t.flock is already set?

stolowski commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -544,13 +568,72 @@ func (t *importTransaction) Cancel() error {
+//
+// The amount of snapshots cleaned is returned or an error
+func CleanupAbandondedImports() (cleaned int, err error) {
+	inProgressSnapshots, err := filepath.Glob(filepath.Join(dirs.SnapshotsDir, "*_importing"))
```
This could go to the "..importing" consts suggested above.

stolowski commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -544,13 +568,72 @@ func (t *importTransaction) Cancel() error {
+// This is meant to be call at startup of snapd before any real imports
+// happen.
+//
+// The amount of snapshots cleaned is returned or an error
```
It seems to actually clean as much as possible and return a number of cleaned snapshots *and* an error (if there was one), but see comment below.

stolowski commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -544,13 +568,72 @@ func (t *importTransaction) Cancel() error {
+		}
+		if err := tr.Cancel(); err != nil {
+			errs = append(errs, err)
+		}
+		cleaned++
```
Should we have 'else' here and only cleaned++ no no error?

stolowski commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -544,13 +568,72 @@ func (t *importTransaction) Cancel() error {
+		for _, err := range errs {
+			fmt.Fprintf(buf, " - %v\n", err)
+		}
+		return cleaned, fmt.Errorf("cannot cleanup imports:\n%s", buf.String())
```
Per above comments this may succeed with some cleanups and fail on others, so maybe "cannot cleanup imports:" if cleaned == len(inProgressSnapshots) and "cannot cleanup some imports:" otherwise?

stolowski commented on overlord/snapshotstate/snapshotmgr.go:
```diff
@@ -58,6 +60,7 @@ type SnapshotManager struct {
 	state *state.State
 
 	lastForgetExpiredSnapshotTime time.Time
+	cleanupAbandondedImportsRan   bool
```
Not needed anymore after moving to StartUp?

mvo5 replied:
Nice catch!

stolowski reviewed (commented):
Looks nice! A few minor remarks.

pedronis commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -486,34 +498,48 @@ var ErrCannotCancel = errors.New("cannot cancel: import already finished")
 type importTransaction struct {
 	id        uint64
+	flockPath string
+	flock     *osutil.FileLock
```
can you explain why you switch to have file locks? because of more than one snapd could be running at  the same time? or something else?

mvo5 replied:
Maybe it's not needed. My thinking was mostly that if something calls "snapshotstate.Import()" while backend.CleanupAbandondedImports() is in progress things would be bad. I don't think this can happen today because we do the cleanup in the StartUp() code so the daemon will not reply to API yet (AIUI). So maybe YAGNI and I could remove the flock. OTOH we still need to keep marker files around so it's not a huge win if we remove the locks (but happy to do it).

mvo5 replied:
The other advantage is that we could call CleanupAbandondedImports periodically - but again, if the exiting code does not have bugs it will never have to cleanup anything so probably not a good reason.

pedronis replied:
I'm not against haveing the locks but as you said that there not clearly currently needed. I asked because we never use file locks far except when interfacting with other procesees. I'm fine if you want to keep them but they probably need a comment like:

```
// flock is a file lock associated with the import, the logic using it prevents potentially cleaning up/cancelling a still in
// progress import in case we started allowing cleaning and imports to happen concurrently
```

pedronis commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -544,13 +573,86 @@ func (t *importTransaction) Cancel() error {
+			errs = append(errs, err)
+			continue
+		}
+		if tr.InProgress() {
```
same question basically, when do we expect this to be true?

pedronis reviewed (commented):
some questions

stolowski commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -475,6 +475,18 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
+var (
+	importingFnRegexp = regexp.MustCompile("^([0-9]+)_importing$")
+	importingFnGlob   = "*_importing"
+	importingFnFmt    = "%d_importing"
```
Thanks for this!

stolowski reviewed (approved):
Looks good, thanks for the changes!

pedronis commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -486,34 +498,52 @@ var ErrCannotCancel = errors.New("cannot cancel: import already finished")
-	return osutil.FileExists(importInProgressFilepath(setID))
+	if flock, err := osutil.OpenExistingLockForReading(t.flockPath); err == nil {
+		defer flock.Close()
+		return flock.TryLock() == osutil.ErrAlreadyLocked
```
given that locks are created first and locked separately, staring at this more this looks racy because is releasing the lock immediately. We probably need to another more global and/or marker values combination for this to work. Also for the use in Iter just checking for existence is still enough I think.

pedronis reviewed (changes requested):
after staring a bit more the lock usage seems racy to me

mvo5 commented:
@pedronis I simplified this and removed the flock(). It still uses the marker files. I also added a comment in CleanupAbandondedImports() that it cannot run concurrently with anything else.

pedronis commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -532,6 +555,9 @@ func (t *importTransaction) Cancel() error {
+		errs = append(errs, err)
+	}
 	if len(errs) > 0 {
 		buf := bytes.NewBuffer(nil)
```
this error formatting seems not tested

pedronis commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -544,13 +570,63 @@ func (t *importTransaction) Cancel() error {
+		}
+	}
+	if len(errs) > 0 {
+		buf := bytes.NewBuffer(nil)
```
this also seems not tested

pedronis commented on overlord/snapshotstate/backend/backend.go:
```diff
@@ -486,34 +498,52 @@ var ErrCannotCancel = errors.New("cannot cancel: import already finished")
+	return false
 }
 
 // Cancel cancels a snapshot import and cleanups any files on disk belonging
```
I don't know if it's needed to comment on this but with the latest changes Cancel can only be called on an  in-progress import. Before I think it would have done nothing instead of erroring as now.

pedronis reviewed (approved):
thanks, some final comments

stolowski commented on overlord/snapshotstate/backend/backend_test.go:
```diff
@@ -944,8 +945,10 @@ func (s *snapshotSuite) TestImport(c *check.C) {
 		importingFile := filepath.Join(dirs.SnapshotsDir, fmt.Sprintf("%d_importing", t.setID))
 		if t.inProgress {
-			err = ioutil.WriteFile(importingFile, nil, 0644)
+			flock, err := osutil.NewFileLock(importingFile)
```
This should be reverted to `err = ioutil.WriteFile(importingFile, nil, 0644)` now that we don't use locks? (not to confuse anyone who may be looking at these tests later).

stolowski commented on overlord/snapshotstate/backend/backend_test.go:
```diff
@@ -1344,11 +1347,53 @@ func (s *snapshotSuite) TestIterWithMockedSnapshotFiles(c *check.C) {
 	callbackCalled = 0
 	fn = "1_importing"
-	err = ioutil.WriteFile(filepath.Join(dirs.SnapshotsDir, fn), nil, 0644)
+	flock, err := osutil.NewFileLock(filepath.Join(dirs.SnapshotsDir, fn))
```
Same.

stolowski reviewed (approved):
Looks good, thank you! Small remark regarding tests that would be nice to clean before merging.

mvo5 commented:
I will merge and do the error testing/cleanup in a followup, will open right away after this is merged
